activation/nipost.go

```diff
@@ -223,41 +222,72 @@ func (nb *NIPostBuilder) BuildNIPost(
 	poetProofDeadline := publishEpochEnd.Add(-nb.poetCfg.CycleGap)
 
 	logger.Info("building nipost",
-		zap.Time("poet round start", poetRoundStart),
-		zap.Time("poet round end", poetRoundEnd),
+		zap.Time("poet round start", curPoetRoundStart),
+		zap.Time("poet round end", curPoetRoundEnd),
 		zap.Time("publish epoch end", publishEpochEnd),
 		zap.Uint32("publish epoch", postChallenge.PublishEpoch.Uint32()),
 	)
 
 	// Phase 0: Submit challenge to PoET services.
-	count, err := nipost.PoetRegistrationCount(nb.localDB, signer.NodeID())
+	addresses := make([]string, 0)
+	for _, client := range nb.poetProvers {
+		addresses = append(addresses, client.Address())
+	}
+
+	// check, if already registered for some PoET services with given addresses
+	existingRegistrations, err := nipost.PoetRegistrations(nb.localDB, signer.NodeID(), addresses...)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get poet registration count: %w", err)
+		return nil, fmt.Errorf("failed to get poet registrations from db: %w", err)
 	}
-	if count == 0 {
-		now := time.Now()
-		// Deadline: start of PoET round for publish epoch. PoET won't accept registrations after that.
-		if poetRoundStart.Before(now) {
-			return nil, fmt.Errorf(
-				"%w: poet round has already started at %s (now: %s)",
-				ErrATXChallengeExpired,
-				poetRoundStart,
-				now,
-			)
-		}
 
-		submitCtx, cancel := context.WithDeadline(ctx, poetRoundStart)
-		defer cancel()
-		err := nb.submitPoetChallenges(submitCtx, signer, poetProofDeadline, challenge.Bytes())
-		if err != nil {
-			return nil, fmt.Errorf("submitting to poets: %w", err)
-		}
-		count, err := nipost.PoetRegistrationCount(nb.localDB, signer.NodeID())
-		if err != nil {
-			return nil, fmt.Errorf("failed to get poet registration count: %w", err)
+	// check if some registrations missing
+	existingRegistrationsMap := make(map[string]bool)
+	for _, er := range existingRegistrations {
+		existingRegistrationsMap[er.Address] = true
+	}
+
+	missingRegistrationsPoETAddresses := make(map[string]bool)
+	for _, addr := range addresses {
+		if _, ok := existingRegistrationsMap[addr]; !ok {
+			missingRegistrationsPoETAddresses[addr] = true
 		}
-		if count == 0 {
-			return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET", source: submitCtx.Err()}
+	}
+
+	if len(missingRegistrationsPoETAddresses) > 0 {
+		now := time.Now()
+
+		// Deadline: start of PoET round: we will not accept registrations after that
+		if !curPoetRoundStart.Before(now) {
```

Comment:
Consider inverting the check and handling the case when it's too late under `if` (return) and the happy path below (without indentation).